csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,290 @@
+
+  void run(TensorView* tv) {
+    // This holds a mapping from scheduled val groups to IterDomains in tv.
+    // Note this is non-const since we insert new IDs into it while replaying.
+    GroupIdMap tv_ids = mapScheduledGroupsToLoopIterDomains(tv);
+
+    // Now, for each ValGroup in abstract, find the closest producer ValGroups
+    // with entries in tv_ids and replay the path from them. If none
+    // exists, then do not include this dimension in the output
+    std::vector<IterDomain*> loop_domain;
+    for (const AbstractId& abs_id : abstract_tensor_.domain) {
+      IterDomain* new_id = replayAbstractId(abs_id, tv_ids);
+      if (new_id == nullptr) {
+        continue;
+      }
+      loop_domain.push_back(new_id);
+    }
+    tv->setLoopDomain(loop_domain);
+  }
+
+  ValGroup abstractIdToValGroup(const AbstractId& abs_id) {
+    NVF_ERROR(
+        abs_id.is<ValGroupAndItsGraph>(),
+        "AbstractId must be a ValGroupAndItsGraph");
+    const ValGroupAndItsGraph& vgg = abs_id.as<ValGroupAndItsGraph>();
+    if (graph_ == nullptr) {
+      graph_ = vgg.graph;
+    } else {
+      NVF_ERROR(graph_ == vgg.graph);
+    }
+    return vgg.group;
+  }
+
+  //! Record all ValGroups that are producers of those in abstract_tensor_.
+  //! This lets us find starting IterDomains for scheduling concrete's loop
+  //! domain.
+  //! TODO: this should probably use a new more general utility that resembles
+  //! StmtSort::getExprsTo in val_graph_visitor.h
+  void findScheduledValGroups() {
+    std::stack<ValGroup> vg_stack;
+    for (const AbstractId& abs_id : abstract_tensor_.domain) {
+      vg_stack.push(abstractIdToValGroup(abs_id));
+    }
+    while (!vg_stack.empty()) {
+      ValGroup vg = vg_stack.top();
+      vg_stack.pop();
+      bool inserted = scheduled_val_groups_.insert(vg).second;
+      if (!inserted) {
+        // Avoid cycles
+        continue;
+      }
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        for (const ValGroup& vg_inp : graph_->inputGroups(eg)) {
+          vg_stack.push(vg_inp);
+        }
+      }
+    }
+  }
+
+  //! Look at all ValGroups that are producers of abstract_tensor_.domain and
+  //! map any of those containing loop IterDomains in tv to those loop
+  //! IterDomains. These provide the starting points for scheduling.
+  GroupIdMap mapScheduledGroupsToLoopIterDomains(TensorView* tv) const {
+    GroupIdMap tv_ids;
+
+    // Look up each loop IterDomain in tv and assert that it is in
+    // scheduled_val_groups_, then map it.
+    for (IterDomain* loop_id : tv->getLoopDomain()) {
+      ValGroup vg = graph_->toGroup(loop_id);
+      // NOTE: this check guarantees that we will only ever _append_ transforms
+      // to the loop domain of tv. If instead we searched for producer IDs in
+      // the root->loop path of tv, then we could wind up silently erasing
+      // some pre-existing transforms.
+      NVF_CHECK(
+          scheduled_val_groups_.count(vg) > 0,
```

Comment:
I didn't expect this should be considered an error. We could do some best-effort propagation (I think that's the behavior of the existing transform propagator). I'm not sure which would be better. Since AbstractTensor could be anything, it may make sense to enforce it must have all corresponding domains, and that may be also good for correctness.

Reply:
> Since AbstractTensor could be anything, it may make sense to enforce it must have all corresponding domains, and that may be also good for correctness.

The line above is checking that the `AbstractTensor` contains all the loop domains of the `TensorView`. Relaxing this requirement is not a bad idea: we would just "ignore" those domains and not schedule them. However, it's unclear where we should place them in the loop domain. For example
```
T0 [ i0 i1 i2 ]
AbstractTensor [ VG{i0 * i2} ]
```
In this case the middle dimension from `T0` is "unscheduled" and the first and last dimensions need to be merged, so where would we put the `i1` dimension in the output? To avoid this ambiguity, I am just forcing us to explicitly include each loop `ValGroup` in the `AbstractTensor` so that this choice is explicit.

Reply:
The existing transform replay also has the same problem. IIRC, it just puts the remaining domains at the end of the list, i.e., the innermost position. 

For AbstractTensor, I think avoiding the ambiguity is better.